feature/amazon/src/main/java/com/revenuecat/purchases/amazon/handler/UserDataHandler.kt

```diff
@@ -50,10 +57,33 @@ class UserDataHandler(
         onError: PurchasesErrorCallback
     ) {
         val userDataRequestId = purchasingServiceProvider.getUserData()
-        requests[userDataRequestId] = onSuccess to onError
+        val request = Request(onSuccess, onError)
+        synchronized(this) {
+            requests[userDataRequestId] = request
+            addTimeoutToUserDataRequest(userDataRequestId)
+        }
     }
 
     private fun PurchasesErrorCallback.invokeWithStoreProblem(message: String) {
         this(PurchasesError(PurchasesErrorCode.StoreProblemError, message))
     }
+
+    private fun addTimeoutToUserDataRequest(requestId: RequestId) {
+        mainHandler.postDelayed(
+            {
+                val request = getRequest(requestId) ?: return@postDelayed
+                val error = PurchasesError(
+                    PurchasesErrorCode.UnknownError,
+                    AmazonStrings.ERROR_TIMEOUT_GETTING_USER_DATA
+                )
+                request.onError(error)
+            },
+            GET_USER_DATA_TIMEOUT_MILLIS
+        )
+    }
+
+    @Synchronized
+    private fun getRequest(requestId: RequestId): Request? {
+        return requests.remove(requestId)
+    }
```

Comment:
Some of this code is very similar or even duplicated between the `ProductDataHandler` and the `UserDataHandler`. I thought about extracting to avoid duplication, but didn't seem to be worth the effort, lmk if you think we should do this.

Reply:
I think it's fine